docs/design/datacontracts/Thread.md

```diff
@@ -0,0 +1,174 @@
+    TS_InMTA                  = 0x00008000,    // Thread is part of the MTA
+
+    // Some bits that only have meaning for reporting the state to clients.
+    TS_ReportDead             = 0x00010000,    // in WaitForOtherThreads()
+    TS_FullyInitialized       = 0x00020000,    // Thread is fully initialized and we are ready to broadcast its existence to external clients
+
+    TS_TaskReset              = 0x00040000,    // The task is reset
+
+    TS_SyncSuspended          = 0x00080000,    // Suspended via WaitSuspendEvent
+    TS_DebugWillSync          = 0x00100000,    // Debugger will wait for this thread to sync
+
+    TS_StackCrawlNeeded       = 0x00200000,    // A stackcrawl is needed on this thread, such as for thread abort
+                                                // See comment for s_pWaitForStackCrawlEvent for reason.
+
+    // unused                 = 0x00400000,
+
+    // unused                 = 0x00800000,
+    TS_TPWorkerThread         = 0x01000000,    // is this a threadpool worker thread?
+
+    TS_Interruptible          = 0x02000000,    // sitting in a Sleep(), Wait(), Join()
+    TS_Interrupted            = 0x04000000,    // was awakened by an interrupt APC. !!! This can be moved to TSNC
+
+    TS_CompletionPortThread   = 0x08000000,    // Completion port thread
+
+    TS_AbortInitiated         = 0x10000000,    // set when abort is begun
+
+    TS_Finalized              = 0x20000000,    // The associated managed Thread object has been finalized.
+                                                // We can clean up the unmanaged part now.
+
+    TS_FailStarted            = 0x40000000,    // The thread fails during startup.
+    TS_Detached               = 0x80000000,    // Thread was detached by DllMain
+}
+
+record struct DacThreadData (
+    uint ThreadId;
+    uint OsThreadId;
+    ThreadState State;
+    bool PreemptiveGCDisabled
+    TargetPointer AllocContextPointer;
+    TargetPointer AllocContextLimit;
+    TargetPointer Frame;
+    TargetPointer FirstNestedException;
+    TargetPointer TEB;
+    TargetPointer LastThrownObjectHandle;
+    TargetPointer NextThread;
+);
+```
+
+## Apis of contract
+``` csharp
+DacThreadStoreData GetThreadStoreData();
+```
+
+## Version 1
+
+
+
+``` csharp
+SListReader ThreadListReader = Contracts.SList.GetReader("Thread");
+
+DacThreadStoreData GetThreadStoreData()
+{
+    TargetPointer threadStore = Target.ReadGlobalTargetPointer("s_pThreadStore");
+    var runtimeThreadStore = new ThreadStore(Target, threadStore);
+
+    TargetPointer firstThread = ThreadListReader.GetHead(runtimeThreadStore.SList.Pointer);
+
+    return new DacThreadStoreData(
+        ThreadCount : runtimeThreadStore.m_ThreadCount,
+        UnstartedThreadCount : runtimeThreadStore.m_UnstartedThreadCount,
+        BackgroundThreadCount : runtimeThreadStore.m_BackgroundThreadCount,
+        PendingThreadCount : runtimeThreadStore.m_PendingThreadCount,
+        DeadThreadCount: runtimeThreadStore.m_DeadThreadCount,
```

Comment:
> UnstartedThreadCount : runtimeThreadStore.m_UnstartedThreadCount,
> BackgroundThreadCount : runtimeThreadStore.m_BackgroundThreadCount,
> PendingThreadCount : runtimeThreadStore.m_PendingThreadCount,
> DeadThreadCount: runtimeThreadStore.m_DeadThreadCount,

These are BCL concepts that we have not moved to C# yet. I am not sure whether it makes sense to have them in the low-level thread contract.

Reply:
I'll split those out into a separate api then. It so happens that `ISOSDacInterface` exposes these to SOS, although I'm not sure how much they are used by SOS clients.

Reply:
Also, these are not currently only BCL concepts, as the iterators we have over threads depend on them. I'm not 100% certain they can just be moved to managed without consequence. It would be nice though.

Reply:
> although I'm not sure how much they are used by SOS clients.

Yeah, I have struggled with random pieces of information returned by DAC SOS APIs many times when doing significant changes in the runtime (and often ended up breaking them in the process without anybody caring).

Reply:
I usually think of our diagnostics in two categories:
- abstractions that are broadly useful for anyone debugging managed code
- abstractions that are only useful for runtime developers trying to understand runtime data structures

I think we have a lot of lattitude to do takebacks on the latter but need to be much more cautious on the former.